### ISSUE TITLE
PDB JSON Schema Only Allows Number

### ISSUE THREAD

**Issue by sstaley-hioscar** [argoproj/argo-workflows#13530]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

I recently deployed argo workflows in my organization's production cluster. I also wrote a wrapper cdk8s chart that engineers and data scientists use to deploy argo workflow charts. In developing the chart, we decided to add json schema validation in order to prevent errant workflow configurations being applied to the cluster.

In doing so we noticed that the pod disruption budget json schema is incorrect. "minAvailable" and "maxAvailable" cannot be set as a percentage for workflows or k8s jobs. They must be set as a numerical value. However, the json schema requires a  string. This allows for "100%", etc, which will not work.
https://raw.githubusercontent.com/argoproj/argo-workflows/v3.5.8/api/jsonschema/schema.json
The offending definition:
```
    "io.k8s.apimachinery.pkg.util.intstr.IntOrString": {
      "type": "string"
    },
```
We changed this to "integer" in our forked schema.

related issues:
https://github.com/argoproj/argo-workflows/issues/6349
https://github.com/argoproj/argo-workflows/issues/6060
https://github.com/argoproj/argo-workflows/issues/4817#issuecomment-754819948

### Version(s)

v3.5.8

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
N/A
```

### Logs from the workflow controller

```text
N/A
```

### Logs from in your workflow's wait container

```text
N/A
```

**Comment by agilgur5** [comment]
This is a follow-up [from Slack](https://cloud-native.slack.com/archives/C01QW9QSSSK/p1722972355571829?thread_ts=1722972355.571829&cid=C01QW9QSSSK). 

> related issues:

I specifically mentioned these 3 that I had stumbled upon that detail dealing with the `IntOrString` variability in OpenAPI v2.
Since k8s supports OpenAPI v2 now (I don't think it did at the time of those issues), this could potentially be resolvable with a schema upgrade. Otherwise, I believe this requires custom tooling to resolve

**Comment by agilgur5** [comment]
I'd been working on https://github.com/argoproj/argo-workflows/issues/12631#issuecomment-1931128000 recently and the parsing difference made me think this might have very similar root cause and symptoms, and indeed it does: the [CRD](https://raw.githubusercontent.com/argoproj/argo-workflows/v3.5.10/manifests/base/crds/full/argoproj.io_workflows.yaml) (search `poddisruptionbudget`) is using an OpenAPI v3 type with `anyOf`, while Argo's API is using OpenAPI v2 which does not have such an option.

This means two things:
- You should be able to submit with an integer fine via k8s
- Since the CRD already has OpenAPI v3, we should be safe to move Argo's API to OpenAPI v3. That would be a breaking change though, require some codegen/tooling modifications, etc

**Comment by sstaley-hioscar** [comment]
> This is a follow-up [from Slack](https://cloud-native.slack.com/archives/C01QW9QSSSK/p1722972355571829?thread_ts=1722972355.571829&cid=C01QW9QSSSK).
> 
> > related issues:
> 
> I specifically mentioned these 3 that I had stumbled upon that detail dealing with the `IntOrString` variability in OpenAPI v2. (credit your source 😉) Since k8s supports OpenAPI v2 now (I don't think it did at the time of those issues), this could potentially be resolvable with a schema upgrade. Otherwise, I believe this requires custom tooling to resolve

Yes, those were from you on slack. Thank you for collecting them. 

By "schema upgrade" Do you mean changing string to integer?

**Comment by sstaley-hioscar** [comment]
If you mean an Open API version upgrade, I disagree. 

The only representation of a PDB that works for argo workflows is an integer. The fact that it _can_ be represented by a string doesn't to me justify making that option available. Are your worried about making a breaking change to the spec?

**Comment by agilgur5** [comment]
> If you mean an Open API version upgrade

> Are your worried about making a breaking change to the spec?

No that's just what the upstream PDB schema is. Argo inherits the `IntOrString` from k8s, but the API tooling is has no idea how to handle it, whereas the CRD tooling is from k8s and so has custom logic for it.

**Comment by sstaley-hioscar** [comment]
Ah, I see. I wouldn't see an issue diverging from the upstream considering the percentage PDB option doesn't apply here. 

I don't think I actually understand the CRD tooling. I'm using the open API spec directly in my cdk8s chart, so I may be out of my depth here. I would just change "string" to "integer" and be done with it.

**Comment by agilgur5** [comment]
> I don't think I actually understand the CRD tooling. I'm using the open API spec directly in my cdk8s chart, so I may be out of my depth here.

So the API's spec is generated from protobufs, whereas the CRD spec is generated directly from Go types, so the tooling pretty much has no overlap as such 😕
I'd like to get rid of protobufs entirely per #11053, as did other contributors, but for now, that is another source of the diff between the API and the CRD.

>  I would just change "string" to "integer" and be done with it.

The entire PDB spec is inherited from k8s; that's where the `io.k8s` part comes from in the schema. 2 levels nested is `IntOrString` with is types the 3rd level of nesting.

>  I wouldn't see an issue diverging from the upstream considering the percentage PDB option doesn't apply here.

Yea you're probably not wrong in this specific case. I'm not sure if Argo has any other use-cases that are string only or would have any.

It's definitely easier to maintain with a straight import/schema inheritance though, so I'm not sure if I'd take that option. And keeps things consistent for users even if they are not _entirely_ identical.

In any case, the `IntOrString` issue affects other fields too, so a wholesale fix for that would be good, which would be the OpenAPI v3 upgrade

**Comment by sstaley-hioscar** [comment]
gRPC is awesome! I'm surprised you'd want to get rid of it. It makes low latency distributed microservices possible for us at Oscar, provides centralized type-checking, and has a much more robust framework for re-using TCP connections/transient error handling. 

But now I understand a bit better why it might be hard to change. Any chance out of curiosity you could point me to where those protobufs are located in the codebase? I was trying to figure out how the open api spec is generated before and got a bit lost.

**Comment by agilgur5** [comment]
You can search by language on GitHub, including protobufs: https://github.com/search?q=repo%3Aargoproj%2Fargo-workflows+language%3A%22Protocol+Buffer%22&type=code

> gRPC is awesome! I'm surprised you'd want to get rid of it.

The core part of Argo is the Controller, which doesn't respond to API requests. The Server is a lot smaller of a piece and is optional as well, but all the codegen around protobufs and maintaining all the tooling and such is not really worth its low usage (and notably, half those deps are deprecated, like gRPC Gateway v1, gogoprotobuf, etc).
There's also some gRPC vs HTTP handling that can occasionally be a pain to debug as well. Notably maintaining a private service is very different from maintaining a compatible OSS API, where HTTP is substantially more common.
Most maintainers have not thought it worth the hassle to maintain, and the deprecated deps mean that removal might be easier than complex upgrades too.

We also already have type-checking and schemas, and Argo doesn't lend itself well to low latency either (neither does the k8s API per se, at least not without potentially causing trouble for other services in the cluster)

But this is getting pretty off-topic 😅
